client/src/use/useModeManager.ts

```diff
@@ -0,0 +1,155 @@
+        Track: {
+            autoAdvanceFrame: boolean;
+        };
+        Detection: {
+            continuous: boolean;
+        };
+    };
+}
+
+export interface Seeker {
+    seek(frame: number): void;
+    nextFrame(): void;
+  }
+/**
+ * The point of this composition function is to define and manage the transition betwee
+ * different UI states within the program.  States and state transitions can be modified
+ * based on settings, blocked if it tries to go to incompatible state or provide feedback
+ *
+ * Mostly allows us to inject additional logic into transitions.
+ */
+export default function useModeManager({
+  selectedTrackId,
+  editingTrack,
+  frame,
+  trackMap,
+  playbackComponent,
+  selectTrack,
+  getTrack,
+  selectNextTrack,
+  addTrack,
+  tsRemoveTrack,
+  setNewDefaultType,
+}: {
+    selectedTrackId: Ref<TrackId | null>;
+    editingTrack: Ref<boolean>;
+    frame: Ref<number>;
+    trackMap: Map<TrackId, Track>;
+    playbackComponent: Ref<Seeker>;
+    selectTrack: (trackId: TrackId | null, edit: boolean) => void;
+    getTrack: (trackId: TrackId) => Track;
+    selectNextTrack: (delta?: number) => TrackId | null;
+    addTrack: (frame: number) => Track;
+    tsRemoveTrack: (trackId: TrackId) => void;
+    setNewDefaultType: (type?: string) => void;
+}) {
+  function handleSelectTrack(trackId: TrackId | null, edit = false) {
+    selectTrack(trackId, edit);
+  }
+  //Handles adding a new track with the NewTrack Settings
+  function handleAddTrack() {
+    selectTrack(addTrack(frame.value).trackId, true);
+  }
+
+  function handleTrackTypeChange({ trackId, value }: { trackId: TrackId; value: string }) {
+    getTrack(trackId).setType(value);
+  }
+  // Default settings which are updated by the CreationMode component
+  // Not making them reactive, and eventually will probably be in localStorage
+  let newTrackSettings: NewTrackSettings |null = null;
+
+  function handleUpdateNewTrackSettings(updatedTrackSettings: NewTrackSettings) {
+    setNewDefaultType(updatedTrackSettings.type);
+    newTrackSettings = updatedTrackSettings;
+  }
+
+  function newTrackSettingsAfterLogic(newTrack: Track) {
+    if (newTrack && newTrackSettings !== null) {
+      if (newTrackSettings.mode === 'Track' && newTrackSettings.modeSettings.Track.autoAdvanceFrame) {
+        playbackComponent.value.nextFrame();
+      } else if (newTrackSettings.mode === 'Detection') {
+        if (newTrackSettings.modeSettings.Detection.continuous) {
+          handleAddTrack();
+        } else { //Deselect the new track
+          selectTrack(newTrack.trackId, false);
+        }
+      }
+    }
+  }
+
+  function handleUpdateRectBounds(frameNum: number, bounds: RectBounds) {
+    if (selectedTrackId.value !== null) {
+      const track = trackMap.get(selectedTrackId.value);
+      if (track) {
+        const features = track.getFeature(frameNum);
+        let newTrack = false;
+        if (!features || features.bounds !== undefined) {
```

Comment:
(2) This line in particular made me a little uneasy, because it leaks implementation details out of Layers.vue.  The presence or absence of `bounds` isn't a good indicator of whether we're creating a new track and should apply the newTrackSettings.

Reply:
Correct me If I am wrong, but I believe that Layers.vue, EditAnnotationLayer.vue don't have any clue that a new Track/Detection is being created explicitly.  EditAnnotationLayer.vue relies on the fact that the there are no bounds for creating a new Rect.  For track/detection creations we just throw it into edit mode without providing bounds.  
To get more information we would have to modify the actual newTrack so that it toggles a flag that it's creating a new track and that if the user exits the creation by clicking on something else or hitting escape it stops the new track (which I can do).

Reply:
> Correct me If I am wrong, but I believe that Layers.vue, EditAnnotationLayer.vue don't have any clue that a new Track/Detection is being created explicitly

No, you're right.  I suppose I should have just said that I didn't like how this was using lack of bounds to be the indicator for "creating new track".  

> To get more information we would have to modify the actual newTrack so that it toggles a flag that it's creating a new track and that if the user exits the creation by clicking on something else or hitting escape it stops the new track (which I can do).

That sounds good to me.

Reply:
There is one small issue here I neglected.  When in Track mode we technically aren't looking for a new track being created, we are looking for a new detection.  Which I believe could only be seen by looking to see if there were existing bounds and then afterwards if we have new bounds.